## ANN - BANK USERS.

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf

In [2]:
df = pd.read_csv("Churn_Modelling.csv")
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [3]:
df.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [5]:
df.isnull()

RowNumber  CustomerId  Surname  CreditScore  Geography  Gender    Age  \
0         False       False    False        False      False   False  False   
1         False       False    False        False      False   False  False   
2         False       False    False        False      False   False  False   
3         False       False    False        False      False   False  False   
4         False       False    False        False      False   False  False   
...         ...         ...      ...          ...        ...     ...    ...   
9995      False       False    False        False      False   False  False   
9996      False       False    False        False      False   False  False   
9997      False       False    False        False      False   False  False   
9998      False       False    False        False      False   False  False   
9999      False       False    False        False      False   False  False   

      Tenure  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      False    False          False      False           False   
1      False    False          False      False           False   
2      False    False          False      False           False   
3      False    False          False      False           False   
4      False    False          False      False           False   
...      ...      ...            ...        ...             ...   
9995   False    False          False      False           False   
9996   False    False          False      False           False   
9997   False    False          False      False           False   
9998   False    False          False      False           False   
9999   False    False          False      False           False   

      EstimatedSalary  Exited  
0               False   False  
1               False   False  
2               False   False  
3               False   False  
4               False   False  
...               ...     ...  
9995            False   False  
9996            False   False  
9997            False   False  
9998            False   False  
9999            False   False  

[10000 rows x 14 columns]

In [6]:
x= df.iloc[:,3:-1].values
y= df.iloc[:,-1].values

In [7]:
print(x)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [8]:
print(y)

[1 0 1 ... 1 1 0]


In [9]:
#Encoding

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
x[:,1]= le.fit_transform(x[:,1])
x[:,2]= le.fit_transform(x[:,2])
x

array([[619, 0, 0, ..., 1, 1, 101348.88],
       [608, 2, 0, ..., 0, 1, 112542.58],
       [502, 0, 0, ..., 1, 0, 113931.57],
       ...,
       [709, 0, 0, ..., 0, 1, 42085.58],
       [772, 1, 1, ..., 1, 0, 92888.52],
       [792, 0, 0, ..., 1, 0, 38190.78]], dtype=object)

In [10]:
# train and test split

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3,random_state = 78)


In [11]:
print(x_train.shape,x_test.shape)

(7000, 10) (3000, 10)


In [12]:
print(y_train.shape,y_test.shape)

(7000,) (3000,)


In [13]:
# Feature Scaling

from sklearn.preprocessing import StandardScaler
ss= StandardScaler()
x_train= ss.fit_transform(x_train)
x_test= ss.fit_transform(x_test)


## Building ANN

In [14]:
# initializing ANN

model = tf.keras.models.Sequential()

In [15]:
# hidden layer 1

model.add(tf.keras.layers.Dense(units=7,activation ="relu"))

In [16]:

model.add(tf.keras.layers.Dense(units=6,activation ="relu"))

In [17]:
model.add(tf.keras.layers.Dense(units=1,activation ="sigmoid"))

# Training the ANN.

In [18]:
model.compile(loss="binary_crossentropy",optimizer='adam',metrics=["accuracy"])

In [19]:
model.fit(x_train,y_train,batch_size =32,epochs=100)

Epoch 1/100
219/219 [==============================] - 1s 2ms/step - loss: 0.6327 - accuracy: 0.6867
Epoch 2/100
219/219 [==============================] - 0s 2ms/step - loss: 0.5146 - accuracy: 0.7987
Epoch 3/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4722 - accuracy: 0.8026
Epoch 4/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4490 - accuracy: 0.8030
Epoch 5/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4365 - accuracy: 0.8076
Epoch 6/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4284 - accuracy: 0.8089
Epoch 7/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4219 - accuracy: 0.8116
Epoch 8/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4147 - accuracy: 0.8151
Epoch 9/100
219/219 [==============================] - 0s 2ms/step - loss: 0.4057 - accuracy: 0.8201
Epoch 10/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3968 - accura

219/219 [==============================] - 0s 2ms/step - loss: 0.3311 - accuracy: 0.8681
Epoch 82/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3315 - accuracy: 0.8651
Epoch 83/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3307 - accuracy: 0.8653
Epoch 84/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3307 - accuracy: 0.8669
Epoch 85/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3311 - accuracy: 0.8664
Epoch 86/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3308 - accuracy: 0.8656
Epoch 87/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3305 - accuracy: 0.8664
Epoch 88/100
219/219 [==============================] - 1s 2ms/step - loss: 0.3302 - accuracy: 0.8669
Epoch 89/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3302 - accuracy: 0.8689
Epoch 90/100
219/219 [==============================] - 0s 2ms/step - loss: 0.3298 - accuracy: 

# Predicting

### QUESTION

when geography is france with credit score 600 and gender male of age 40 of tenure 3 whose balance is 60000 with number of product 2 and has a credit card of active member with estimated salary $ 50000
 
Find should we say goodbye to the customer? 


In [40]:
model.predict(ss.transform([[0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]))> 0.5


1/1 [==============================] - 0s 21ms/step


array([[False]])

In [41]:
# test set result.
y_pred = model.predict(x_test)
y_pred = (y_pred > 0.5)

model_predict = (y_pred,y_test)



94/94 [==============================] - 0s 1ms/step


In [42]:
y_pred

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

In [43]:
y_test

array([0, 0, 0, ..., 0, 1, 0], dtype=int64)

In [44]:
print(np.concatenate((y_pred.reshape(len(y_pred),1),y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 1]
 [0 0]]


In [45]:
# confusion Matrix


from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test,y_pred)
ac_sc=accuracy_score(y_test,y_pred)
print(cm)
print("Accuracy_score:",round(ac_sc,4))




[[2290   88]
 [ 339  283]]
Accuracy_score: 0.8577


## Conclusion.

The accuray of the Artificial neural network model is 86% . Therefore the model performs good on the bank user dataset.